### PR TITLE
Toggling other fields

### PR DESCRIPTION
**Pull request by pun-ky**

### CODE REVIEW

File: src/main/kotlin/com/neva/gradle/fork/config/Config.kt [hunk 1]
```diff
@@ -28,18 +29,7 @@ abstract class Config(private val fork: ForkExtension, val name: String) {
...
-      prompts.forEach { name, prompt ->
-        val definition = fork.propertyDefinitions.get(prompt.name) ?: PropertyDefinition(prompt.name)
-        val property = Property(others, definition, prompt)
-
-        others[name] = property
-      }
-
-      return others.values.toList()
-    }
+  val definedProperties: List<Property> by lazy { propsDefine() }
```
Comment (pun-ky) [inline src/main/kotlin/com/neva/gradle/fork/config/Config.kt]: `Property` needs to hold `enabled` state

File: src/main/kotlin/com/neva/gradle/fork/config/Config.kt [hunk 2]
```diff
@@ -102,6 +92,20 @@ abstract class Config(private val fork: ForkExtension, val name: String) {
...
+    val context = PropertyContext(others)
+
+    prompts.forEach { name, prompt ->
+      val definition = fork.propertyDefinitions.get(prompt.name) ?: PropertyDefinition(prompt.name)
+      val property = Property(definition, prompt, context)
+
+      others[name] = property
+    }
+
+    return others.values.sortedBy { p -> fork.propertyDefinitions.indexOf(p.name) }.toList()
```
Comment (pun-ky) [inline src/main/kotlin/com/neva/gradle/fork/config/Config.kt]: now order is deterministic (reflected order in `properties { define { } }` block

File: src/main/kotlin/com/neva/gradle/fork/config/Config.kt [hunk 3]
```diff
@@ -135,17 +139,21 @@ abstract class Config(private val fork: ForkExtension, val name: String) {
       return
     }
 
-    val fileProps = Properties()
-    fileProps.load(FileInputStream(file))
-    fileProps.forEach { p, v -> prompts[p.toString()]?.value = v.toString() }
+    FileInputStream(file).use { input ->
```
Comment (pun-ky) [inline src/main/kotlin/com/neva/gradle/fork/config/Config.kt]: closeable not used to that it was not possible to delete `build/fork/config/properties.properties` etc

File: src/main/kotlin/com/neva/gradle/fork/config/properties/Property.kt [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
 
+/**
+ * Represents prompted property bound with its definition.
+ * Can interact with other properties using common context.
+ */
 class Property(
-  private val others: Map<String, Property>,
-  private val definition: PropertyDefinition,
-  private val prompt: PropertyPrompt
+  val definition: PropertyDefinition,
```
Comment (pun-ky) [inline src/main/kotlin/com/neva/gradle/fork/config/properties/Property.kt]: definition potentially usable while implementing `control { other("propName").definition.type == /* */ } `

File: src/main/kotlin/com/neva/gradle/fork/gui/PropertyDialogField.kt [hunk 5]
```diff
@@ -62,7 +65,7 @@ class PropertyDialogField(
 
   companion object {
 
-    val DESCRIPTION_TEXT_COLOR = Color(128, 128, 128)
+    val DESCRIPTION_TEXT_COLOR = Color(0, 0, 128)
```
Comment (pun-ky) [inline src/main/kotlin/com/neva/gradle/fork/gui/PropertyDialogField.kt]: to distinguish disabled fields and descriptions

File: src/main/kotlin/com/neva/gradle/fork/config/Config.kt (hunk 2 above)
Comment (mierzwid) [reply src/main/kotlin/com/neva/gradle/fork/config/Config.kt]: like

File: src/main/kotlin/com/neva/gradle/fork/config/properties/Property.kt [hunk 6]
```diff
@@ -31,9 +35,44 @@ class Property(
   val invalid: Boolean
     get() = validate().hasErrors()
 
-  fun validate(): PropertyValidator = PropertyValidator(this).apply(definition.validator)
+  var enabled: Boolean = true
```
Comment (mierzwid) [inline src/main/kotlin/com/neva/gradle/fork/config/properties/Property.kt]: redundant type declaration ;)

File: src/main/kotlin/com/neva/gradle/fork/config/properties/Property.kt [hunk 7]
```diff
@@ -31,9 +35,44 @@ class Property(
...
 
-  fun other(name: String): Property {
-    return others[name] ?: throw PropertyException("Property named '$name' is not defined.")
+  var required: Boolean = definition.required
```
Comment (mierzwid) [inline src/main/kotlin/com/neva/gradle/fork/config/properties/Property.kt]: same

File: src/main/kotlin/com/neva/gradle/fork/gui/PropertyDialog.kt [hunk 8]
```diff
@@ -127,6 +133,7 @@ class PropertyDialog(private val config: Config) {
 
   private fun validateAllFields() {
     fields.forEach(PropertyDialogField::assignValue)
+    fields.forEach(PropertyDialogField::control)
```
Comment (mierzwid) [inline src/main/kotlin/com/neva/gradle/fork/gui/PropertyDialog.kt]: like

File: src/main/kotlin/com/neva/gradle/fork/config/properties/PropertyDefinition.kt [hunk 9]
```diff
@@ -16,11 +19,9 @@ open class PropertyDefinition @Inject constructor(val name: String) {
...
   var required = true
 
-  var validator: PropertyValidator.() -> Unit = {
-    if (required) {
-      required()
-    }
-  }
+  var controller: Property.() -> Unit = {}
+
+  var validator: PropertyValidator.() -> Unit = { notBlank() }
```
Comment (mierzwid) [inline src/main/kotlin/com/neva/gradle/fork/config/properties/PropertyDefinition.kt]: Having field optional when I will only override a validator seems to me unpleasant surprise

Comment (pun-ky) [reply src/main/kotlin/com/neva/gradle/fork/config/properties/PropertyDefinition.kt]: nope, this is fixed in `Property` / recovered to your version

**Pull request merged**